src/modules/endpoint/sagas/watchFetchCategory.js

```diff
@@ -9,32 +9,66 @@ import type {
 } from '../../app/StoreActionType'
 import type { DataContainer } from '../DataContainer'
 import loadCityContent from './loadCityContent'
+import { ContentLoadCriterion } from '../ContentLoadCriterion'
+import DatabaseContext from '../DatabaseContext'
+
+function * getRootAvailableLanguages (
+  context: DatabaseContext,
+  loadCriterion: ContentLoadCriterion, dataContainer: DataContainer): Saga<Map<string, string>> {
+
+  if (loadCriterion.shouldLoadLanguages()) {
+    const languages = yield call(dataContainer.getLanguages, context)
+    return new Map<string, string>(languages
+      .map(language => [language.code, `/${context.cityCode}/${language.code}`]))
+  }
+
+  // If there are no loaded languages the result is an empty map because we do not have a root category
+  return new Map<string, string>()
+}
 
 function * fetchCategory (dataContainer: DataContainer, action: FetchCategoryActionType): Saga<void> {
-  const {city, language, path, depth, key, forceUpdate, shouldRefreshResources} = action.params
+  const { city, language, path, depth, key, criterion } = action.params
   try {
-    yield call(loadCityContent, dataContainer, city, language, forceUpdate, shouldRefreshResources)
+    const loadCriterion = new ContentLoadCriterion(criterion)
+    const cityContentLoaded = yield call(loadCityContent, dataContainer, city, language, loadCriterion)
 
-    const [categoriesMap, resourceCache, languages] = yield all([
-      call(dataContainer.getCategoriesMap),
-      call(dataContainer.getResourceCache),
-      call(dataContainer.getLanguages)
-    ])
+    if (cityContentLoaded) {
+      // Only proceed if the content is ready to be pushed to the state. If not then the UI automatically displays an
+      // appropriate error
 
-    const insert: PushCategoryActionType = {
-      type: `PUSH_CATEGORY`,
-      params: {
-        categoriesMap,
-        resourceCache,
-        path,
-        languages,
-        depth,
-        key,
-        city,
-        language
+      const context = new DatabaseContext(city, language)
+      const [categoriesMap] = yield all([
+        call(dataContainer.getCategoriesMap, context)
+      ])
+
+      let resourceCache = {}
+
+      const resourceCacheAvailable = yield call({context: dataContainer, fn: dataContainer.resourceCacheAvailable}, context)
+
+      if (resourceCacheAvailable) {
+        // TODO: This call should happen paralel
```

Comment:
Shouldn't we create an issue or do it in this task?

Reply:
Fixed this